app/src/main/java/io/appium/uiautomator2/handler/TouchEvent.java

```diff
@@ -0,0 +1,73 @@
+package io.appium.uiautomator2.handler;
+
+import android.graphics.Rect;
+import android.support.test.uiautomator.UiObjectNotFoundException;
+
+import com.jayway.jsonpath.JsonPath;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import io.appium.uiautomator2.handler.request.SafeRequestHandler;
+import io.appium.uiautomator2.http.AppiumResponse;
+import io.appium.uiautomator2.http.IHttpRequest;
+import io.appium.uiautomator2.model.AndroidElement;
+import io.appium.uiautomator2.model.KnownElements;
+import io.appium.uiautomator2.server.WDStatus;
+import io.appium.uiautomator2.utils.Logger;
+
+public abstract class TouchEvent extends SafeRequestHandler {
+    protected int clickX, clickY;
+    private AndroidElement element;
+    private boolean isElement;
+
+    public TouchEvent(String mappedUri) {
+        super(mappedUri);
+    }
+
+    @Override
+    public AppiumResponse safeHandle(IHttpRequest request) {
+        String path = "$.params.";
+        try {
+            JSONObject json = getPayload(request);
+            if (json.has("id")) {
+                String id = json.getString("id");
+                element = KnownElements.getElementFromCache(id);
+
+                if (element != null)
+                    isElement = true;
+                if (isElement) {
```

Comment:
Is it possible to get here and not have an element? With the logic as it stands, if we get to here and there is no element, we will not populate `clickX` and `clickY` at all. Not sure what the consequences of that would be, though it seems like the touch event would not work.